### DISCUSSION TITLE
Can I use asyncio with Dagster?

### DISCUSSION THREAD

**Discussion by sryza** [dagster-io/dagster#16269] [Q&A] [answered]
Can I use use asnycio for concurrency within an asset / op? What about across ops?

**Comment by sryza** [comment, marked as answer]
When defining an asset or op, you can write an async def, e.g.:

```python
@asset
async def asset1():
    ...
```


This allows using asyncio for concurrency *within* the asset / op.

We don't support asyncio for concurrency across asset / ops.  Here's an issue where we're tracking the request to add this: https://github.com/dagster-io/dagster/issues/4041.  Using aiohttp within an asset/op is a great fit if you need to make N concurrent web fetches for your calculations.

If you are using an async def with the the default multi-process executor you will end up with process-level concurrency across ops / assets and the ability to use asnycio for concurrency within the assets/ops.

**Reply by whillas-yabble** [reply]
Hey,
I'm trying to get this working but keep getting back coroutines instead awaited coroutines? Is there an example of this working? Perhaps its because I'm trying to run this in a test using `execute_in_process`? If so is there an example of a test for async Ops?

**Reply by alangenfeld** [reply]
You can reference test coverage for examples https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster_tests/core_tests/test_op_invocation.py

If you are directly invoking the op / asset you will have to await the coroutine. `execute_in_process` on a job/graph containing the asset/op should work.
